runtime/lua/vim/filetype.lua

```diff
@@ -2449,6 +2562,56 @@ local function match_pattern(name, path, tail, pat)
   return (tail:match(fullpat))
 end
 
+--- Cache for whether current candidates match fast patterns.
+--- Need to be manually reset in every new `M.match()` call.
+--- @type table<string,boolean>
+local fast_matches = {}
+
+--- @param name string
+--- @param path string
+--- @param tail string
+--- @param pat string
+--- @return boolean
+local function match_fastpat(name, path, tail, pat)
```

Comment:
Can this be combined with `match_pattern`?

Unless you can show otherwise, I don't think the additional code is worth the perf gain. And in general I'm not sure we need to introduce a new "fastpat" concept.

Reply:
It can be combined, but behind an extra flag argument of whether it is regular or fast pattern. The reason for this is that regular filetype matching [depends on whether pattern contains slash or not](https://github.com/neovim/neovim/blob/028dd3c5c4d1828eec64c099d3372ffb90572dc0/runtime/lua/vim/filetype.lua#L2444). If reused direcrly for fast patterns, it would require both types of patterns to share this property, which limits the choice of good fast patterns and needs to be documented (as fragile detail). For example, using 'utt' fast pattern for '.*/%.mutt/muttrc.*' filetype pattern.

Hence there is a separate pattern which tries to match all three possible candidates, independently of slash presence.

Reply:
> I'm not sure we need to introduce a new "fastpat" concept.

Apart from mentioned more general matching logic, the main reason for "fast pattern" as a concept was that their matching result is cached (hence the name). It is harmful to cache regular filetype matching as it is used at most once.

The general "parse and generate tree" which works like nested `if ... then ... else` *and* also respects priority sorting can indeed remove the need for caching. However it feels like having small extra reward for a single filetype matching stuff compared to the complexity it entails.

Reply:
Consider adding an extra flag for now then. I think the whole slash behaviour needs to be revisited as, like priority, is adding significant complexity.

Reply:
I've added an extra `try_all_candidates` flag with a comment. No more extra function.